File: bindings/csharp/Everanium.LibItb3.Tests/RuntimeTests.cs
// Runtime diagnostics surface: GOMAXPROCS query / set / restore, the
// heap-profile writer, the pool-counter snapshot and its slot layout,
// and the hash-registry enumeration.

namespace Everanium.Itb3.Tests;

public class RuntimeTests
{
    [Fact]
    public void GomaxprocsQuerySetRestore()
    {
        int orig = Runtime.SetGOMAXPROCS(0);
        Assert.True(orig > 0);
        Assert.Equal(orig, Runtime.SetGOMAXPROCS(-3));
        Assert.Equal(orig, Runtime.SetGOMAXPROCS(orig + 1));
        Assert.Equal(orig + 1, Runtime.SetGOMAXPROCS(0));
        Assert.Equal(orig + 1, Runtime.SetGOMAXPROCS(orig));
    }

    [Fact]
    public void HeapProfileWrittenAndEmptyPathRejected()
    {
        string dir = Path.Combine(
            Path.GetTempPath(),
            $"itb-loop-test-heap-{Environment.ProcessId}");
        Directory.CreateDirectory(dir);
        string path = Path.Combine(dir, "heap.prof");
        Runtime.WriteHeapProfile(path);
        Assert.True(new FileInfo(path).Length > 0);
        Directory.Delete(dir, recursive: true);

        // The empty path falls back to ITB_MEMPROFILE inside libitb3;
        // with the variable clear there is nothing to fall back to.
        Environment.SetEnvironmentVariable("ITB_MEMPROFILE", null);
        var ex = Assert.Throws<ItbException>(() => Runtime.WriteHeapProfile(""));
        Assert.Equal(Status.BadInput, ex.Status);
    }

    [Fact]
    public void PoolStatsLayout()
    {
        int len = Runtime.PoolStatsLen();
        Assert.True(len >= 9);
        long[] v = Runtime.PoolStats();
        Assert.Equal(len, v.Length);
        long tiers = v[0];
        Assert.True(tiers > 0);
        Assert.Equal(1 + 5 * tiers + 8, len);
    }

    [Fact]
    public void HashNamesCanonical()
    {
        string[] names = Pipeline.HashNames();
        Assert.Equal("aesitb128", names[0]);
        Assert.Contains("areion512", names);
    }
}

File: bindings/vbnet/tests/Everanium.LibItb3.VisualBasic.Tests/RuntimeTests.vb
' Runtime diagnostics surface: GOMAXPROCS query / set / restore, the
' heap-profile writer, the pool-counter snapshot and its slot layout,
' and the hash-registry enumeration.

Imports System.IO
Imports Everanium.Itb3.VisualBasic
Imports Xunit

Public Class RuntimeTests

    <Fact>
    Public Sub GomaxprocsQuerySetRestore()
        Dim orig As Integer = Library.SetGOMAXPROCS(0)
        Assert.True(orig > 0)
        Assert.Equal(orig, Library.SetGOMAXPROCS(-3))
        Assert.Equal(orig, Library.SetGOMAXPROCS(orig + 1))
        Assert.Equal(orig + 1, Library.SetGOMAXPROCS(0))
        Assert.Equal(orig + 1, Library.SetGOMAXPROCS(orig))
    End Sub

    <Fact>
    Public Sub HeapProfileWrittenAndEmptyPathRejected()
        Dim dir As String = Path.Combine(
            Path.GetTempPath(), $"itb-loop-test-heap-vbnet-{Environment.ProcessId}")
        Directory.CreateDirectory(dir)
        Dim profilePath As String = Path.Combine(dir, "heap.prof")
        Library.WriteHeapProfile(profilePath)
        Assert.True(New FileInfo(profilePath).Length > 0)
        Directory.Delete(dir, True)

        ' The empty path falls back to ITB_MEMPROFILE inside libitb3;
        ' with the variable clear there is nothing to fall back to.
        Environment.SetEnvironmentVariable("ITB_MEMPROFILE", Nothing)
        Dim ex As ItbException = Assert.Throws(Of ItbException)(
            Sub() Library.WriteHeapProfile(""))
        Assert.Equal(Status.BadInput, ex.Status)
    End Sub

    <Fact>
    Public Sub PoolStatsLayout()
        Dim len As Integer = Library.PoolStatsLen()
        Assert.True(len >= 9)
        Dim v As Long() = Library.PoolStats()
        Assert.Equal(len, v.Length)
        Dim tiers As Long = v(0)
        Assert.True(tiers > 0)
        Assert.Equal(CLng(len), 1L + 5L * tiers + 8L)
    End Sub

    <Fact>
    Public Sub HashNamesCanonical()
        Dim names As String() = Pipeline.HashNames()
        Assert.Equal("aesitb128", names(0))
        Assert.Contains("areion512", names)
    End Sub
End Class
